convex/lib/withSession.ts

```diff
@@ -34,47 +39,29 @@ const optionalSessionIdValidator = v.union(v.id("sessions"), v.null());
  * @param func - Your function that can take in a `session` in the first (ctx) param.
  * @returns A function to be passed to `query` or `mutation`.
  */
-export function withOptionalSession<
-  Ctx extends QueryCtx,
-  ArgsValidator extends PropertyValidators,
-  Output
->(
-  fn: ValidatedFunction<
-    Ctx & { session: Doc<"sessions"> | null },
-    ArgsValidator,
-    Promise<Output>
-  >
-): ValidatedFunction<
-  Ctx,
-  ArgsValidator & { sessionId: typeof optionalSessionIdValidator },
-  Promise<Output>
->;
-export function withOptionalSession<Ctx extends QueryCtx, Output>(
-  fn: UnvalidatedFunction<
-    Ctx & { session: Doc<"sessions"> | null },
-    [],
-    Promise<Output>
-  >
-): ValidatedFunction<
-  Ctx,
-  { sessionId: typeof optionalSessionIdValidator },
-  Promise<Output>
->;
-export function withOptionalSession(fn: any) {
-  const handler = fn.handler ?? fn;
-  const args = fn.args ?? {};
-  return {
-    args: {
-      ...args,
-      sessionId: optionalSessionIdValidator,
-    },
-    handler: async (ctx: any, allArgs: any) => {
-      const { sessionId, ...args } = allArgs;
-      const session = sessionId ? await ctx.db.get(sessionId) : null;
-      return handler({ ...ctx, session }, args);
-    },
-  };
-}
+const optionalSessionMiddlewareValidator = {
+  sessionId: v.union(v.null(), v.id("sessions")),
+};
+const transformContextForOptionalSession = async <Ctx>(
+  ctx: Ctx & { db: DatabaseReader },
+  args: ObjectType<typeof optionalSessionMiddlewareValidator>
+): Promise<Ctx & { session: Doc<"sessions"> | null }> => {
+  const session = args.sessionId ? await ctx.db.get(args.sessionId) : null;
```

Comment:
can we do this to be more bulletproof, especially for people upgrading to 0.17.0 for an existing app
```suggestion
  const sessionId = args.sessionId ? ctx.db.normalizeId("sessions", args.sessionId) : null;
  const session = sessionId ? await ctx.db.get(args.sessionId) : null;
```

Reply:
Hmm I wonder if we should do a separate function for people upgrading then. This change alone wouldn't be enough (the validator would fail first), and there's other things one might want for upgrading (e.g. writing back the updated value to local storage / session storage)